File: Java/35.search-insert-position.java
public class Solution {
    public int searchInsert(int[] nums, int target) {
        if (target <= nums[0]) {
            return 0;
        }
        if (target > nums[nums.length - 1]) {
            return nums.length;
        }
        return search(nums, target, 0, nums.length - 1);
    }

    private int search(int[] nums, int target, int startIndex, int endIndex) {
        if (startIndex == endIndex) {
            return startIndex;
        }
        int middle = (startIndex + endIndex) / 2;
        if (target == nums[middle]) {
            return middle;
        }
        if (target < nums[middle]) {
            return search(nums, target, startIndex, middle);
        }
        return search(nums, target, middle + 1, endIndex);
    }
}

File: 35.search-insert-position.js
/*
 * @lc app=leetcode id=35 lang=javascript
 *
 * [35] Search Insert Position
 */

// @lc code=start
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number}
 */

// Binary Search O(logN) O(1)
var searchInsert = function (nums, target) {
  let pivot,
    left = 0,
    right = nums.length - 1;
  while (left <= right) {
    pivot = left + parseInt((right - left) / 2);
    if (nums[pivot] == target) return pivot;
    if (target < nums[pivot]) right = pivot - 1;
    else left = pivot + 1;
  }

  return left;
};
// @lc code=end
